dask/array/core.py

```diff
@@ -1938,14 +1998,96 @@ def normalize_chunks(chunks, shape=None):
 
     return tuple(tuple(int(x) if not math.isnan(x) else x for x in c) for c in chunks)
 
 
+def auto_chunks(chunks, shape, limit, dtype, previous_chunks=None):
+    """ Determine automatic chunks
+
+    See also
+    --------
+    normalize_chunks: for full docstring and parameters
+    """
+    chunks = list(chunks)
+
+    autos = {i for i, c in enumerate(chunks) if c == 'auto'}
+    if not autos:
+        return tuple(chunks)
+
+    if limit is None:
+        limit = config.get('array.chunk-size')
+        if isinstance(limit, str):
+            limit = parse_bytes(limit)
+
+    if dtype is None:
+        raise TypeError("DType must be known for auto-chunking")
+
+    if dtype.hasobject:
+        raise NotImplementedError(
+            "Can not use auto rechunking with object dtype. "
+            "We are unable to estimate the size in bytes of object data")
+
+    limit = max(1, limit // dtype.itemsize)
+
+    largest_block = np.prod([cs if isinstance(cs, Number) else max(cs)
+                             for cs in chunks if cs != 'auto'])
+
+    if previous_chunks:
+        result = {a: np.median(previous_chunks[a]) for a in autos}
+        multiplier = limit / largest_block / np.prod(list(result.values()))
+        last = 0
+        while (multiplier, autos) != last:  # while things improve
+            last = (multiplier, set(autos))  # record last value for stopping condition
+            for a in sorted(autos):
+                proposed = result[a] * multiplier ** (1 / len(autos))
+                if proposed > shape[a]:  # we've hit the shape boundary
+                    autos.remove(a)
+                    largest_block *= shape[a]
+                    chunks[a] = shape[a]
+                    del result[a]
+                else:
+                    result[a] = round_to(proposed, shape[a])
+
+            # recompute how much multiplier we have left, repeat
+            multiplier = limit / largest_block / np.prod(list(result.values()))
+
+        for k, v in result.items():
+            chunks[k] = v
+        return tuple(chunks)
+
+    else:
+        size = (limit / largest_block) ** (1 / len(autos))
+        small = [i for i in autos if shape[i] < size]
+        if small:
+            for i in small:
+                chunks[i] = (shape[i],)
+            return auto_chunks(chunks, shape, limit, dtype)
+
+        for i in autos:
+            chunks[i] = round_to(size, shape[i])
+
+        return tuple(chunks)
+
+
+def round_to(c, s):
+    """ Return a chunk dimension that is close to an even multiple
+
+    We want the largest factor of the dimension size (s) that is less than the
+    desired chunk size, but not less than half, which is too much.  If no such
+    factor exists then we just go with the original chunk size and accept an
+    uneven chunk at the end.
+    """
+    try:
+        return max(f for f in factors(s) if c / 2 <= f <= c)
+    except ValueError:
```

Comment:
This corresponds to no matching factors? (probably worth a comment)

Reply:
Added a comment